streaming/streaming.go

```diff
@@ -0,0 +1,233 @@
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"time"
+
+	dataflow "cloud.google.com/go/dataflow/apiv1beta3"
+	datastream "cloud.google.com/go/datastream/apiv1alpha1"
+	datastreampb "google.golang.org/genproto/googleapis/cloud/datastream/v1alpha1"
+	dataflowpb "google.golang.org/genproto/googleapis/dataflow/v1beta3"
+	fieldmaskpb "google.golang.org/protobuf/types/known/fieldmaskpb"
+
+	"github.com/cloudspannerecosystem/harbourbridge/common/utils"
+	"github.com/cloudspannerecosystem/harbourbridge/profiles"
+)
+
+type SrcConnCfg struct {
+	Name     string
+	Location string
+}
+
+type DstConnCfg struct {
+	Name     string
+	Location string
+	Prefix   string
+}
+
+type DatastreamCfg struct {
+	StreamId                    string
+	StreamLocation              string
+	StreamDisplayName           string
+	SourceConnectionConfig      SrcConnCfg
+	DestinationConnectionConfig DstConnCfg
+}
+
+type DataflowCfg struct {
+	JobName  string
+	Location string
+}
+
+type StreamingCfg struct {
+	DatastreamCfg DatastreamCfg
+	DataflowCfg   DataflowCfg
+}
+
+// VerifyAndUpdateCfg checks the fields and errors out if certain fields are empty.
+// It then auto-populates certain empty fields like StreamId and Dataflow JobName.
+func VerifyAndUpdateCfg(streamingCfg *StreamingCfg, dbName string) error {
+	dsCfg := streamingCfg.DatastreamCfg
+	if streamingCfg.DatastreamCfg.StreamLocation == "" {
+		return fmt.Errorf("please specify DatastreamCfg.StreamLocation in the streaming config")
+	}
+	srcCfg := dsCfg.SourceConnectionConfig
+	if srcCfg.Name == "" || srcCfg.Location == "" {
+		return fmt.Errorf("please specify Name and Location under DatastreamCfg.SourceConnectionConfig in the streaming config")
+	}
+	dstCfg := dsCfg.DestinationConnectionConfig
+	if dstCfg.Name == "" || dstCfg.Location == "" {
+		return fmt.Errorf("please specify Name and Location under DatastreamCfg.DestinationConnectionConfig in the streaming config")
+	}
+
+	dfCfg := streamingCfg.DataflowCfg
+	if dfCfg.Location == "" {
+		return fmt.Errorf("please specify the Location under DataflowCfg in the streaming config")
+	}
+
+	// If both ID and Display name are empty, generate a new one for both.
+	// If either is present, assign it to the other one.
+	if dsCfg.StreamId == "" && dsCfg.StreamDisplayName == "" {
+		// TODO: Update names to have more info like dbname.
+		streamId, err := utils.GenerateName("hb-stream-" + dbName)
+		if err != nil {
+			return fmt.Errorf("error generating stream name: %v", err)
+		}
+		streamingCfg.DatastreamCfg.StreamId = streamId
+		streamingCfg.DatastreamCfg.StreamDisplayName = streamId
+	} else if dsCfg.StreamId == "" {
+		streamingCfg.DatastreamCfg.StreamId = streamingCfg.DatastreamCfg.StreamDisplayName
+	} else if dsCfg.StreamDisplayName == "" {
+		streamingCfg.DatastreamCfg.StreamDisplayName = streamingCfg.DatastreamCfg.StreamId
+	}
+
+	if dfCfg.JobName == "" {
+		// Update names to have more info like dbname.
+		jobName, err := utils.GenerateName("hb-dataflow" + dbName)
```

Comment:
should the jobName have prefix as "hb-dataflow-"? Is the hyphen absent at the end deliberate ?